### PR TITLE
Added Shodan

### PR DESCRIPTION
**Pull request by rachejazz**
Bypassing cloudflare security check still left.
Should I use [cloudscraper](https://pypi.org/project/cloudscraper/) ?

### CODE REVIEW

File: core/util/shodan.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+
+import re
+import os
```
Comment (saeeddhqan) [inline core/util/shodan.py]: remove it

File: core/util/shodan.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+class main:
+
+	def __init__(self, framework, q, key, limit=1, count=10):
```
Comment (saeeddhqan) [inline core/util/shodan.py]: limit and count didn't use in the code

File: core/util/shodan.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+			framework  : core attribute
+			q          : query for search
+			key		   : API key
+			ip		   : query on a given host IP
+			limit      : count of pages
+		"""
+		self.framework = framework
+		self.q = q
+		self.key = key
```
Comment (saeeddhqan) [inline core/util/shodan.py]: remove it

File: core/util/shodan.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+		self._pages = ''
+		self.limit = limit
```
Comment (saeeddhqan) [inline core/util/shodan.py]: remove it

File: core/util/shodan.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+		self.num = count
```
Comment (saeeddhqan) [inline core/util/shodan.py]: remove it

File: core/util/shodan.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+		self.shodan_api = f"https://api.shodan.io/shodan/host/search?key={self.key}&query={self.q}"
+		self._links = []
```
Comment (saeeddhqan) [inline core/util/shodan.py:35]: you didn't use it so remove it as well

File: core/util/shodan.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+	def run_crawl(self):
+		self.framework.verbose('[SHODAN] Searching in shodan...')
+		try:
+			req = self.framework.request(self.shodan_api)
+			print(req.text)
```
Comment (saeeddhqan) [inline core/util/shodan.py]: remove it

File: core/util/shodan.py [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+			self.framework.verbose('[SHODAN] request was not accepted!')
+		else:
+			self.acceptable = True
+
+	@property
+	def pages(self):
+		return self._pages
+
+	@property
+	def links(self):
```
Comment (saeeddhqan) [inline core/util/shodan.py:58]: remove links function

File: core/util/shodan.py [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+		return self._links
+
+	@property
+	def dns(self):
+		return self.framework.page_parse(self._pages).get_dns(self.q)
+
+	@property
+	def docs(self):
```
Comment (saeeddhqan) [inline core/util/shodan.py]: remove it. Because shodan is not used as a source for files

File: modules/search/shodan.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+		'name': 'Shodan Search',
+		'author': 'Divya Goswami',
+		'version': '1.0',
+		'description': 'Search your query on shodan.io and show the results.',
+		'sources': ('shodan',),
+		'options': (
+			('query', None, True, 'Query string', '-q', 'store'),
+			('limit', 1, False, 'Search limit(number of pages, default=1)', '-l', 'store'),
+			('key', None, False, 'shodan.io api key [required]', '-k', 'store'),
+			('count', 50, False, 'Number of results per page(min=10, max=100, default=50)', '-c', 'store'),
```
Comment (saeeddhqan) [inline modules/search/shodan.py]: you didn't use limit and count in shodan class so remove these vars completely

File: modules/search/shodan.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+			('output', False, False, 'Save output to workspace', '--output', 'store_true'),
+		),
+        'examples': ('shodan -q <QUERY> -k <KEY> -l 15 --output',)
+	}
+
+	def module_run(self):
```
Comment (saeeddhqan) [inline modules/search/shodan.py:40]: you didn't return any value from the shodan class.

**Comment by saeeddhqan** [comment]
Bypassing cloudflare security check still left.
    Why is it important?

File: core/util/shodan.py (hunk 6 above)
Comment (rachejazz) [reply core/util/shodan.py:35]: Api key is required for accessing searches. How do I remove it?
I will remove limits for sure, thank you. Although count is needed. I will update in sometime.

**Comment by rachejazz** [comment]
> Why is it important?
```
    <h1><span data-translate="checking_browser">Checking your browser before accessing</span> shodan.io.</h1>
    
    <div id="no-cookie-warning" class="cookie-warning" data-translate="turn_on_cookies" style="display:none">
      <p data-translate="turn_on_cookies" style="color:#bd2426;">Please enable Cookies and reload the page.</p>
    </div>
    <p data-translate="process_is_automatic">This process is automatic. Your browser will redirect to your requested content shortly.</p>
    <p data-translate="allow_5_secs" id="cf-spinner-allow-5-secs" >Please allow up to 5 seconds&hellip;</p>
    <p data-translate="redirecting" id="cf-spinner-redirecting" style="display:none">Redirecting&hellip;</p>
  </div>
   ...
   truncated
```

Browser check blocking search while using requests.

**Comment by saeeddhqan** [comment]
Really? did you use the API and still have this error?

**Comment by rachejazz** [comment]
> Really? did you use the API and still have this error?

Yep :( the simple command I used:
```
shodan -q ship -k [API] 
```

**Comment by saeeddhqan** [comment]
I think something is wrong!! the API, as the name implies, is used in the applications. Not in the browsers.
Could you send your URL in the slack?

**Comment by rachejazz** [comment]
Yep, I have shared it!

**Comment by rachejazz** [comment]
[asciicast](https://asciinema.org/a/QWad72Wtk4aMhY5DLXbW9LSQQ)

Please check the following module. API issue is now working. (Key displayed is already revoked :D)
Shodan this time works on both search queries and host IP search through DNS!

Waiting for the new core update. I am hoping API modules(such as this one) would not then need to be loaded by default. I will get an if case implemented for all future API modules to download dependencies from marketplace for it to work.

**Comment by saeeddhqan** [comment]
Work on the new version.

**Comment by rachejazz** [comment]
Working with new v2. Closing

**Pull request closed without merging**